hoomd/hpmc/ShapeMoves.h

```diff
@@ -0,0 +1,870 @@
+            {
+            Scalar a = fmax(-stepsize, -(m_params[type_id][i]));
+            Scalar b = fmin(stepsize, (1.0 - m_params[type_id][i]));
+            hoomd::UniformDistribution<Scalar> uniform(a, b);
+            Scalar r = hoomd::detail::generate_canonical<double>(rng);
+            Scalar x = (r < this->m_move_probability) ? uniform(rng) : 0.0;
+            m_params[type_id][i] += x;
+            }
+        pybind11::object d = m_python_callback(type_id, m_params[type_id]);
+        pybind11::dict shape_dict = pybind11::cast<pybind11::dict>(d);
+        shape = typename Shape::param_type(shape_dict);
+        detail::MassProperties<Shape> mp(shape);
+        this->m_det_inertia_tensor = mp.getDetInertiaTensor();
+        }
+
+    void retreat(uint64_t timestep)
+        {
+        // move has been rejected.
+        std::swap(m_params, m_params_backup);
+        }
+
+    pybind11::list getParams(std::string typ)
+        {
+        unsigned int type_id = this->getValidateType(typ);
+        pybind11::list ret;
+        for (unsigned int i = 0; i < m_params[type_id].size(); i++)
+            {
+            ret.append(m_params[type_id][i]);
+            }
+        return ret;
+        }
+
+    void setParams(std::string typ, pybind11::list params)
+        {
+        unsigned int type_id = this->getValidateType(typ);
+        auto N = pybind11::len(params);
+        m_params[type_id].resize(N);
+        m_params_backup[type_id].resize(N);
+        for (unsigned int i = 0; i < N; i++)
+            {
+            m_params[type_id][i] = params[i].cast<Scalar>();
+            m_params_backup[type_id][i] = params[i].cast<Scalar>();
+            }
+        }
+
+    pybind11::dict getTypeParams()
+        {
+        pybind11::dict ret;
+        for (unsigned int type_id = 0; type_id < this->m_ntypes; type_id++)
+            {
+            std::string type_name = this->m_sysdef->getParticleData()->getNameByType(type_id);
+            pybind11::list l;
+            for (unsigned int i = 0; i < m_params[type_id].size(); i++)
+                {
+                l.append(m_params[type_id][i]);
+                }
+            ret[type_name.c_str()] = l;
+            }
+        return ret;
+        }
+
+    Scalar getParamMoveProbability()
+        {
+        return this->m_move_probability;
+        }
+
+    void setParamMoveProbability(Scalar select_ratio)
+        {
+        this->m_move_probability = fmin(select_ratio, 1.0);
+        }
+
+    pybind11::object getCallback()
+        {
+        return m_python_callback;
+        }
+
+    void setCallback(pybind11::object python_callback)
+        {
+        m_python_callback = python_callback;
+        }
+
+    private:
+    unsigned int m_num_params;                        // cache the number of parameters.
+    std::vector<std::vector<Scalar>> m_params_backup; // all params are from 0,1
+    std::vector<std::vector<Scalar>> m_params;        // all params are from 0,1
```

Comment:
The comments don't give information on what the member variables are for though.